#  THE USE OF AI IN SUSTAINABLE AGRICULTURE



### Importing  the relevant libraries

In [1]:
import warnings
warnings.filterwarnings('ignore')
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
#from scipy.stats import randint
import os,io

In [2]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.svm import SVR
svr_rbf = SVR(kernel="rbf", C=100, gamma=0.1, epsilon=0.1)
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn import metrics
import sklearn.model_selection as selection

In [3]:
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

In [4]:
import ipywidgets
from  ipywidgets import interact  

## Understanding the dataset

In [51]:
# checking the working directory 
%pwd

'C:\\Users\\funmi\\Downloads\\A_DEPLOYMENT'

### Getting the data

In [5]:
rabbit_dung_df_original = pd.read_csv(r"C:\Users\funmi\Downloads\PORTFOLIO\Build_Portforlio\Real_Brd_New\Data2_Rabbit_organic.csv", delimiter=";")

In [6]:
rabbit_dung_df_original

Experiments  Stem_lenght  Root_lenght  Stem_diameter  Chlorophyll  \
0   CK_Cla                 0.00         1.00           0.08         1.00   
1     T10_ Cla             0.22         0.21           0.00         0.67   
2     T20_ Cla             0.48         0.56           0.71         0.58   
3     T30_ Cla             0.50         0.53           0.54         0.17   
4     T40_ Cla             1.00         0.20           0.85         0.00   
5     T50_ Cla             0.84         0.00           1.00         0.67   
6          CK_Cuc          0.00         0.48           0.74         0.04   
7       T10_ Cuc           0.47         1.00           0.37         0.00   
8       T20_ Cuc           0.38         0.72           0.35         0.37   
9       T30_ Cuc           0.71         0.76           0.66         0.12   
10      T40_ Cuc           1.00         0.36           1.00         0.00   
11      T50_ Cuc           0.99         0.00           0.95         1.00   

    Above_Ground_Dry_Weight  Above_Ground_Fresh_Weight     N     P    Ca  \
0                      0.00                       0.00  0.00  0.00  0.38   
1                      0.00                       0.22  0.18  1.00  1.00   
2                      0.50                       0.83  0.49  0.54  0.84   
3                      0.50                       0.61  1.00  0.37  0.38   
4                      0.50                       0.83  0.53  0.27  0.22   
5                      1.00                       1.00  0.89  0.22  0.00   
6                      0.00                       0.50  0.00  0.75  0.00   
7                      0.46                       0.38  0.61  0.49  1.00   
8                      0.00                       0.11  0.06  0.85  0.75   
9                      0.44                       0.55  0.17  1.00  0.55   
10                     0.94                       0.97  0.00  0.09  0.22   
11                     1.00                       1.00  1.00  0.48  0.00   

      Mg   CEI  
0   0.00  0.25  
1   0.75  0.47  
2   1.00  0.65  
3   0.84  0.55  
4   1.00  0.54  
5   0.98  0.66  
6   0.25  0.25  
7   0.52  0.53  
8   0.64  0.42  
9   0.72  0.56  
10  0.76  0.53  
11  1.00  0.74

### Exploratory Data Analysis

In [7]:
## checking ths shape of the dataset 
rabbit_dung_df_original.shape

(12, 12)

### Data Description
Experiment: name of stage of experiment

Stem_lenght

Root_lenght

Sten_Diameter

Chlorophy

Above_Ground_Dry_Weight

Above_Ground_Fresh_Weight

N,P,Ca,Mg,CEI : Nitrogen ratio,Pottasium ratio,Calcium ratio,Magnessium ratio, and Compehensive Evaluation Index respectively. 



## Simulated Data

In [8]:
  rabbit_dung_df= pd.read_csv(r"C:\Users\funmi\Downloads\A_DEPLOYMENT\simulated_data_5000.csv")

In [9]:
rabbit_dung_df.tail()

Experiments  Stem_lenght  Root_lenght  Stem_diameter  Chlorophyll  \
4995         NaN       0.4838     0.401320       0.504792     0.750159   
4996         NaN       0.4946     0.628213       0.691856     0.733028   
4997         NaN       0.4939     0.595482       0.538485     0.042748   
4998         NaN       0.4080     0.400305       0.314403     0.147185   
4999         NaN       0.3028     0.273903       0.553710     0.244513   

      Above_Ground_Dry_Weight  Above_Ground_Fresh_Weight         N         P  \
4995                 0.803295                   0.923341  0.497189  0.193838   
4996                 0.193532                   0.981649  0.178419  0.096073   
4997                 0.681312                   0.358199  0.935674  0.186790   
4998                 0.820922                   0.448770  0.572957  0.546553   
4999                 0.588836                   0.431935  0.039418  0.733986   

            Ca        Mg       CEI  
4995  0.764421  0.125575  0.343556  
4996  0.664050  0.910000  0.631395  
4997  0.302116  0.504855  0.463990  
4998  0.575489  0.704840  0.774247  
4999  0.274666  0.920194  0.062362

### Checking for missing values 


In [10]:
rabbit_dung_df.isnull().sum()

Experiments                  4988
Stem_lenght                     0
Root_lenght                     0
Stem_diameter                   0
Chlorophyll                     0
Above_Ground_Dry_Weight         0
Above_Ground_Fresh_Weight       0
N                               0
P                               0
Ca                              0
Mg                              0
CEI                             0
dtype: int64

### Observation
Experiment column would be dropped due to large sum of missing values in the simulated column
also because it is not so relevant to the analysis.

In [11]:
## Count the different experimental stages in the da
rabbit_dung_df.drop('Experiments', axis= 1, inplace = True)

In [12]:
print("average Calcium in Mixture is: {0:.2f}".format(rabbit_dung_df['Ca'].mean()))

average Calcium in Mixture is: 0.50


In [13]:
print("average Pottassium in Mixture is: {0:.2f}".format(rabbit_dung_df['P'].mean()))

average Pottassium in Mixture is: 0.50


In [14]:
print("average Nitrogen in Mixture is: {0:.2f}".format(rabbit_dung_df['N'].mean()))

average Nitrogen in Mixture is: 0.51


In [15]:
print('Nitrogen Basic Statistics')
print("minimum Nitrogen in Mixture is: {0:.2f}".format(rabbit_dung_df['N'].min()))
print("average Nitrogen in Mixture is: {0:.2f}".format(rabbit_dung_df['N'].mean()))
print("maximum Nitrogen in Mixture is: {0:.2f}".format(rabbit_dung_df['N'].max()))

Nitrogen Basic Statistics
average Nitrogen in Mixture is: 0.00
average Nitrogen in Mixture is: 0.51
average Nitrogen in Mixture is: 1.00


In [16]:
rabbit_dung_df.describe()

Stem_lenght  Root_lenght  Stem_diameter  Chlorophyll  \
count  5000.000000  5000.000000    5000.000000  5000.000000   
mean      0.468733     0.495199       0.488710     0.502905   
std       0.102189     0.288784       0.290866     0.287527   
min       0.000000     0.000000       0.000000     0.000000   
25%       0.396100     0.246524       0.232718     0.250600   
50%       0.463500     0.493057       0.487380     0.503181   
75%       0.534025     0.746323       0.737592     0.749385   
max       1.000000     1.000000       1.000000     1.000000   

       Above_Ground_Dry_Weight  Above_Ground_Fresh_Weight            N  \
count              5000.000000                5000.000000  5000.000000   
mean                  0.503842                   0.499219     0.507342   
std                   0.289206                   0.287520     0.290693   
min                   0.000000                   0.000000     0.000000   
25%                   0.253490                   0.252700     0.256733   
50%                   0.506115                   0.498957     0.505965   
75%                   0.752646                   0.749392     0.765514   
max                   1.000000                   1.000000     1.000000   

                 P           Ca           Mg          CEI  
count  5000.000000  5000.000000  5000.000000  5000.000000  
mean      0.498400     0.496164     0.502509     0.500402  
std       0.286308     0.289632     0.287363     0.586987  
min       0.000000     0.000000     0.000000    -0.509569  
25%       0.252665     0.248691     0.252202    -0.009757  
50%       0.500205     0.493253     0.505117     0.493655  
75%       0.741419     0.751066     0.748822     0.996588  
max       1.000000     1.000000     1.000000     1.529945

### Building the Predictive Modelling 


In [17]:
### checking columns that are categorical
cat_var=[item for item in rabbit_dung_df.columns if rabbit_dung_df[item].dtype =='object']
cat_var

[]

In [22]:
### Cehcking the columns
rabbit_dung_df.columns

Index(['Stem_lenght', 'Root_lenght', 'Stem_diameter', 'Chlorophyll',
       'Above_Ground_Dry_Weight', 'Above_Ground_Fresh_Weight', 'N', 'P', 'Ca',
       'Mg', 'CEI'],
      dtype='object')

In [18]:
## check data type
rabbit_dung_df.dtypes

Stem_lenght                  float64
Root_lenght                  float64
Stem_diameter                float64
Chlorophyll                  float64
Above_Ground_Dry_Weight      float64
Above_Ground_Fresh_Weight    float64
N                            float64
P                            float64
Ca                           float64
Mg                           float64
CEI                          float64
dtype: object

In [19]:
rabbit_dung_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Stem_lenght                5000 non-null   float64
 1   Root_lenght                5000 non-null   float64
 2   Stem_diameter              5000 non-null   float64
 3   Chlorophyll                5000 non-null   float64
 4   Above_Ground_Dry_Weight    5000 non-null   float64
 5   Above_Ground_Fresh_Weight  5000 non-null   float64
 6   N                          5000 non-null   float64
 7   P                          5000 non-null   float64
 8   Ca                         5000 non-null   float64
 9   Mg                         5000 non-null   float64
 10  CEI                        5000 non-null   float64
dtypes: float64(11)
memory usage: 429.8 KB


In [21]:
## Checking the target variabble
len(rabbit_dung_df['Stem_lenght'].unique())

4267

Observation:
    The unique values in stem lenght are about 4267
All the data types are now floats. 

In [23]:
## Create target variable 
y = rabbit_dung_df['Stem_lenght']

In [24]:
x= rabbit_dung_df.drop(['Stem_lenght'],axis =1)

In [25]:
x.shape

(5000, 10)

In [26]:
y.shape

(5000,)

### Splitting the dataset

In [32]:
X_train, X_test, Y_train, Y_test= train_test_split(x, y, test_size=0.2, random_state=0)

In [33]:
## build the model
model = RandomForestRegressor()
model.fit(X_train,Y_train)


RandomForestRegressor()

In [34]:
Rand_y_pred = model.predict(X_test)

## Evaluate the Model

In [35]:
rnd_acc =model.score(X_test,Y_test)
rnd_acc

0.9438919931613157

In [39]:
print('RMSE',np.sqrt(metrics.mean_squared_error(Y_test,Rand_y_pred)))

RMSE 0.023811255225229113


In [36]:
from sklearn.svm import SVR
svr_rbf = SVR(kernel="rbf", C=100, gamma=0.1, epsilon=0.1)


In [37]:
svr_rbf.fit(X_train,Y_train)

SVR(C=100, gamma=0.1)

In [38]:
svr_acc =svr_rbf.score(X_test,Y_test)
svr_acc

0.7092336997852837

### Saving the Model

In [40]:
import pickle

In [41]:
## create the file name
Rabbitfile = open('model.pkl','wb')

In [42]:
## save it using pickle and open the file and write binary
pickle.dump(model,Rabbitfile)

In [43]:
save_model = pickle.load(open('model.pkl','rb'))


In [44]:
save_model.predict(X_test)

array([0.400798, 0.398513, 0.474661, 0.378427, 0.452152, 0.55465 ,
       0.525688, 0.52121 , 0.322221, 0.385895, 0.400843, 0.381644,
       0.458175, 0.594731, 0.353435, 0.547544, 0.411715, 0.405721,
       0.430075, 0.507398, 0.394267, 0.444336, 0.358041, 0.357272,
       0.585709, 0.405576, 0.354914, 0.477838, 0.431376, 0.550042,
       0.367535, 0.488351, 0.603054, 0.272374, 0.485266, 0.395627,
       0.370245, 0.505345, 0.335771, 0.417746, 0.567073, 0.642763,
       0.372373, 0.43781 , 0.514094, 0.524559, 0.300559, 0.513362,
       0.560847, 0.54901 , 0.41353 , 0.473753, 0.507383, 0.478001,
       0.467955, 0.381191, 0.432017, 0.48584 , 0.513404, 0.486638,
       0.359766, 0.526622, 0.37161 , 0.468384, 0.421806, 0.351558,
       0.411786, 0.623754, 0.368377, 0.460714, 0.338195, 0.581118,
       0.422469, 0.508517, 0.450625, 0.485831, 0.471916, 0.519801,
       0.504986, 0.406425, 0.390024, 0.554905, 0.513265, 0.314835,
       0.61236 , 0.596106, 0.465805, 0.422144, 0.437447, 0.517

In [45]:
X_test.head()

Root_lenght  Stem_diameter  Chlorophyll  Above_Ground_Dry_Weight  \
398      0.814671       0.611831     0.373458                 0.255336   
3833     0.852527       0.561988     0.662811                 0.648905   
4836     0.964869       0.183070     0.115421                 0.387099   
4572     0.748341       0.105396     0.363035                 0.263967   
636      0.919793       0.094849     0.544126                 0.740510   

      Above_Ground_Fresh_Weight         N         P        Ca        Mg  \
398                    0.105874  0.354898  0.413173  0.675823  0.658160   
3833                   0.363982  0.075991  0.807706  0.289716  0.848930   
4836                   0.393811  0.806253  0.468889  0.062461  0.920668   
4572                   0.538239  0.061416  0.762464  0.106527  0.271402   
636                    0.756927  0.606716  0.166787  0.797421  0.720111   

           CEI  
398  -0.367126  
3833  0.731925  
4836  0.251404  
4572  1.087228  
636  -0.334417

In [46]:
## Make stem lenght preddictions
save_model.predict([[1,0.48,0.74,0.04,0.0,0.50,0.00,0.75,0.00,0.25]])

array([0.31024])